### DISCUSSION TITLE
Service discovery for MCPs

### DISCUSSION THREAD

**Discussion by positiveblue** [modelcontextprotocol/modelcontextprotocol#69] [Ideas - General]
First of all thank you very much for developing and open sourcing this protocol. I think this is a significant step forward for creating interoperable systems, particularly for AI agents interacting with/consuming external tools.

## Current state and limitations
In the current state MCP provides:
  - A standardized framework for integrating external tools and data sources with LLMs.
  - Robust features for managing context, resources, tools, and prompts.

However, there is **no built-in mechanism for service discovery** which means that agents are limited to interacting only with services they are already aware of.

We have been working for a while on a protocol that overlaps with MCP and we solved that limitation using a custom URI that the agent can use to identify compatible services, retrieve their metadata, and seamlessly integrate with them.

That would allow flows like pasting a URI in a chat and having the LLM decide if it needs to integrate with that tool, when to use it etc... (ofc, security is important here so things like SSL may be critical here). You can see an [example of it working here](https://www.youtube.com/watch?v=eRQc6_zcuLM).

## Proposed Enhancement

Introduce a custom URI scheme, such as `mcp://`, to enable:
 1) **Service Identification**: AI agents can distinguish MCP-compatible services from others.
 2) **Metadata Access**: Agents can retrieve metadata (e.g., service type, documentation, endpoints) about the service.
 3) **Dynamic Integration**: Use metadata to dynamically connect to and interact with services.

Example Workflow:
1.	**Discovery**: An AI agent encounters a custom URI, e.g., `mcp://api.myservice.com`.

2.	**Metadata Retrieval**: The agent resolves the URI and fetches detailed information about the service. For example, it performs an HTTP GET request to a predefined endpoint such as: `https://api.myservice.com/llms.txt.`
3.	**Metadata Structure**: The service responds with a JSON or text file containing all relevant metadata, including:
	•	*Authentication*: Supported methods for authenticating requests (e.g., API keys, OAuth).
	•	*Description*: Overview of the service and its capabilities.
	•	*Supported Features*: List of features/tools/resources offered by the service.
	•	*API Documentation*: Link to full API documentation for deeper integrations.
	•	*Payment Information*: Details about pricing, payment methods, or supported payment protocols.

4.	**Integration**: Using the retrieved metadata, the agent:
	•	Configures the necessary authentication and permissions.
	•	Maps supported features to its workflow or capabilities.
	•	Initiates interaction with the service, such as executing tools or subscribing to resources.
  
I would love to hear your thoughts on this and am happy to contribute further to its development if people find it useful.

**Comment by jspahrsummers** [comment]
Possibly related: https://github.com/orgs/modelcontextprotocol/discussions/84

**Comment by displague** [comment]
In an agentic framework, it seems limiting to offer static definitions. A dynamic definition exchange could look like:
1. Discovery: api.example.com/.well-known/mcp (or stateless via adapters to convert OpenAPI to MCP and well known locations for OpenAPI spec: https://github.com/OAI/OpenAPI-Specification/issues/2128)
2. Query: api.example.com/discovery-endpoint?q=(query). Query parameters and patterns could be structured, partial, or anything LLM parseable
3. Query Response:
    a. Match: Responses: (Strong) api.example.com/existing-endpoint with signature, (weak) .../related-endpoint with signature
    b. **Creation**: Response: (Strong) api.example.com/dynamically-created-endpoint with signature

For dynamic endpoints, some control mechanisms may be needed to cancel or persist the endpoint, iterate on the definition, or modify or expand upon implementation details. The latter could result in the server changing the code of the dynamically generated endpoint or removing ambiguity based on user (agent) conversations.

**Reply by jspahrsummers** [reply]
It feels like 2 and 3 should exist within the protocol itself, as these mechanisms already exist. Or where they don't, it seems like a change that should be made to the protocol spec, rather than introducing more out of band communication.

Agreed about 1. See https://github.com/orgs/modelcontextprotocol/discussions/84 capturing a similar idea.

**Comment by YE-YI7** [comment]
Great thread — service discovery is clearly a critical next step for MCP. I want to add a dimension that hasn't been discussed yet: value metadata

The gap
The proposals here focus on *capability discovery* — what a server can do, how to authenticate, what tools/resources it exposes. This is essential. But when an agent faces multiple servers that offer the same capability (e.g., three different image generation APIs), it still can't make an informed choice because there's no structured data about:
Pricing — How much does it cost? Per-token? Per-image? Per-second? Tiered?
Quality— How good is the output? What benchmark scores? Self-reported or third-party verified?
SLA— What's the p50/p99 latency? Uptime? Rate limits?
Payment— API key? Stripe? AP2?
Without this, discovery only answers "what exists" — not "what's worth using."

What I've been working on
I've been designing Agent Service Manifest (ASM) — a lightweight JSON Schema extension that lets providers declare standardized value descriptors alongside their MCP tool definitions. Think of it as: `.well-known/mcp` tells you *what a server can do*, ASM tells you *what a server is worth*.
Key design decisions (aligned with MCP principles):
Composable, not category-specific — one schema covers LLM, image gen, video gen, TTS, GPU compute, etc.
All fields optional — providers expose what they can. Only 3 required fields: `asm_version`, `service_id`, `taxonomy`
`extensions` namespace— category-specific data (like `max_resolution` for image gen) lives in extensions, keeping the core schema stable
Zero breaking changes— can ship as `x-asm` annotations that non-supporting clients simply ignore

What exists so far:
- Schema v0.2 (JSON Schema, 5 top-level modules: pricing, quality, SLA, payment, extensions)
- 18-category taxonomy covering LLM/vision/video/audio/code/data/infra
- 3 real-world demos with actual pricing data: Claude Sonnet 4, FLUX 1.1 Pro, Google Veo 3.1
- Minimal scoring function: user preferences → normalized scores → ranked list + reasoning

Real-world validation:AWS recently shipped an [AWS Marketplace MCP Server](https://docs.aws.amazon.com/marketplace/latest/APIReference/marketplace-mcp-server.html) that does product comparison and recommendation through MCP. This proves the demand — but it's a closed, platform-locked implementation. ASM aims to provide the same capability as an open standard.

How this connects to service discovery
I think the discovery flow should be:
1. .well-known/mcp → capability metadata (what the server can do)
2. x-asm annotations → value metadata (what the server is worth)
3. Agent preference function → ranked selection + reasoning
4. AP2 / payment → execute transaction

Step 1 is what this thread is designing. Step 2 is the gap ASM fills. Steps 3-4 complete the autonomous service selection loop.
Questions for the community
1. Does adding value metadata (pricing/quality/SLA) feel like it belongs in the service discovery mechanism, or should it be a separate layer?
2. Would it make sense to define a standard `x-asm` namespace in ToolAnnotations, or is a separate metadata endpoint better?
3. I'm also in conversation with the [Agent Receipts](https://github.com/agent-receipts/ar) team about pairing ASM (pre-service declaration) with signed receipts (post-service verification) for a complete trust loop. Would this kind of interop be interesting to the MCP community?
Happy to share the full schema, demo code, or a detailed writeup. Here's the [schema overview and 3 cross-category demos](link-to-your-repo-or-doc) if anyone wants to dig in.

**Comment by kinthaiofficial** [comment]
Service discovery for MCPs is a genuinely hard problem because of the decentralized nature of the ecosystem. A few approaches worth considering:

**Static registry approach:**
Central registry (like npm/PyPI but for MCP servers). Servers register their `agent-skills.json` (or equivalent capability manifest). Clients query the registry to find servers that offer a given capability. Simple but requires a trusted central authority.

**DNS-based discovery:**
Similar to service mesh DNS. MCPs register at `_mcp._tcp.domain.com`. Clients resolve via mDNS for local discovery, standard DNS for remote. Decentralized but requires infrastructure setup.

**Well-known endpoint:**
`/.well-known/mcp-capabilities` on any domain that runs an MCP server, similar to `/.well-known/openid-configuration`. Clients that know a domain can auto-discover capabilities without a central registry.

**The A2A Protocol approach:**
Google's Agent-to-Agent protocol proposes `/.well-known/agent.json` for agent capability advertisement. A similar pattern for MCP servers would make sense — `/.well-known/mcp.json` describing what tools are available, what auth is required, what cost estimates look like.

**Federation:**
Multiple registries that sync with each other (like ActivityPub's federated social network model). No single point of failure, decentralized authority.

**What I'd prioritize:**
The well-known endpoint approach first — it's the lowest overhead for existing MCP server operators to adopt. Then standardize the capability manifest schema. Registry/discovery layer can come after the manifest format is stable.

The agent routing use case — "I need to find an MCP server that can do X, what's the cheapest/fastest option?" — requires both capability description AND cost/latency metadata in the manifest.

**Comment by sauravsingla** [comment]
An interesting distinction here may be between discovering available capabilities and deciding which discovered capabilities should enter model context.

We’ve been exploring the second problem in AgentWeave. The architecture keeps a larger capability catalog outside the LLM and performs deterministic task-specific routing before inference.

This suggests a possible two-stage architecture:

service discovery → capability catalog → task-specific routing → model-visible capabilities.

I’d be interested in whether others see discovery and inference-time selection as separate protocol/client responsibilities.

Paper: https://arxiv.org/abs/2608.23078
Code: https://github.com/sauravsingla/agentweave

**Comment by YE-YI7** [comment]
@sauravsingla Yes — I think these should be separate responsibilities. Discovery can enumerate stable capability descriptors and locate related metadata; the client or harness can then construct the smaller model-visible set for the current task.

There is one additional boundary that seems worth making explicit: routing often needs mutable, source-linked facts — cost, SLA, setup, policy/risk, and evidence freshness — without treating those facts as part of the tool schema or immutable capability identity. A pricing or reachability update should be able to change the routing decision without redefining the MCP tool.

Since my April comment above, implementation work has pushed me toward sidecar/registry placement rather than adding these facts to ToolAnnotations. In ASM, external sidecar facts are carried in manifests (the DeepSeek Harness fixture calls them Selection Facts), and an unsigned Selection Receipt v0.1 pins the candidate set, constraints, manifest digests, and selected service. The receipt is audit/provenance only; it does not execute or authorize the selected tool.

AgentWeave already implements the client-side half of this boundary: its MCP example routes plain `tools/list` descriptors before inference, while Issue #22 calls for explicit routing provenance. Would you be open to reviewing one small external fixture: two MCP descriptors plus two ASM sidecars, routed under one task constraint to one model-visible tool, with the candidates, constraint, exact manifest digests, and selected service recorded in a Selection Receipt? I can prepare it outside AgentWeave first; no MCP core-field or AgentWeave adoption claim would be implied.

**Reply by sauravsingla** [reply]
Yes, I’d be happy to review it.

The separation you describe makes sense to me: MCP tool descriptors can remain stable capability identities, while mutable selection facts such as cost, SLA, reachability, policy/risk, and evidence freshness live outside the tool schema and influence routing at runtime.

A small external fixture with two MCP descriptors, two ASM sidecars, one task constraint, and a Selection Receipt sounds like a good way to test that boundary without implying any MCP core-field change or AgentWeave adoption.

Please go ahead and prepare it outside AgentWeave first. I’d be glad to review the fixture, the routing decision, and whether the receipt captures enough provenance to make the selection independently inspectable and reproducible.

**Comment by YE-YI7** [comment]
@sauravsingla Thanks — the external fixture is ready for review: https://github.com/YE-YI7/asm-spec/pull/16

It keeps the two MCP descriptors unchanged, binds each to one synthetic ASM sidecar, applies one explicit structured constraint, selects one model-visible descriptor, and pins both manifest digests plus the decision in Selection Receipt v0.1.

It deliberately does not generate AgentWeave native `ToolRoutingProvenance` or execute/authorize either tool. The fixture is pinned to AgentWeave commit `1f2e9c8`.

Reproduce with:

```bash
python3 tools/build_agentweave_selection_fixture.py --check
python3 -m pytest tools/asm-gen/test_agentweave_selection_fixture.py
```

Three review questions:

1. Is the provisional sidecar-to-descriptor binding at the right boundary?
2. Does the receipt capture enough pre-inference provenance?
3. Should AgentWeave native provenance eventually reference the receipt by digest, URI plus digest, or neither?

The fixture PR is stacked on the selector-hardening PR https://github.com/YE-YI7/asm-spec/pull/15, and no AgentWeave adoption is implied.

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
